KEP: Consistent Reads from Cache

jpbetz opened the pull request:
This idea has been around for awhile and there is interest in working on it for 1.18, so I've put it in KEP form.

cc @wojtek-t @jingyih @liggitt

enj commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+
+## Proposal
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
```
Would it be possible for the api server to discover the version of etcd on start up?  If the feature goes GA and is enabled by default, it would be nice to not require config changes to use it (i.e. upgrading etcd would be enough).

wojtek-t replied:
We've briefly discussed this with @jpbetz yesterday. 
How I would see that is:
- introduce a mechanism to control switching it off by a flag (I'm not 100% convinced it should be that flag, but that's a minor detail)
- as you described, periodically detect etcd version and use it to automatically disable the feature if etcd version is too old
- the flag will just be a safety mechanism for the operator, rather than something everyone has to explicitly set

With that, by default operators won't have to care about that at all, but they will have a mechanism to disable just in case.
Assuming the above, I'm slightly concerned with reusing the `--storage-backend` flag to and using it to control multiple different things. But i agree it's not an issue for now, so maybe that's fine.

jpbetz replied:
> > I took a cursory glance a question came to my mind. How would this work for paginated list calls? I couldn't find anything in the proposal on that.
> 
> It's not solved. That said, I think that we don't necessary need to address that from day1.
> In my opinion, the biggest gain from it we can get from node-originating requests (e.g. kubelet listing pods scheduled on its node). For those requests, the size of the response is small (it fits a single page, assuming you won't make it extremely small), whereas the number of objects to process is proportional to cluster-size (so fairly big).
> 
> So as an example, in 5k-node cluster with 30pods/node, listing "pods from node X" returns 30 items, but listing from etcd would result in listing 150k pods, deserializing them, filtering and returning those 30. Listing from cache means just filtering and returning those 30, and once [kubernetes/kubernetes#85445](https://github.com/kubernetes/kubernetes/pull/85445) merges (and one more small PR on top of it), it will be proportional to the size of result.

This is a valuable insight. I've added it to the motivation section. (and added to wojtek-t to KEP authors).

> 
> So yes - this doesn't immediately solve all problem, but it may visibly help with many of those.
> 
> I can imagine the algorithm like:
> 
> * if limit is set, check if we set selectors - if not fallback to etcd, if so, assume high probability the result will not be too large
> * ensure cache is up-to-date
> * try to list from cache
> * if result is too large, fallback to list from etcd
> 
> Joe Betz - while the above algorithm is probably not something we want to put in the KEP (do we?), the above explanation I made actually is probably worth putting (parts of it should do into motivation, I think).

Humm, so the problem is that, because the watch cache does not paginate responses, clients that were previously getting pagination for RV="" requests will start getting paginated responses.

I'll comment about this and add it to the list of things we'd need to address before going to beta.

If it turns out to be a real issue, maybe it would best be solved by supporting pagination in the watch cache? I haven't though about the complications around that...

enj commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability).
```
Is this related to https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md?

wojtek-t replied:
Yes - that is the api for it.
Currently the implementation is that we serve a bookmark only short before watch would timeout. It solves 99% of cases, but we can be smarter than that.

jpbetz replied:
Added link.

jingyih commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+etcd, since watch cache progress would already kept largely up-to-date
+by the progress events etcd proactively sent.
+
+Optional: For some or all of the etcd progress watch events, also create a
```
Does this require changes to etcd watch API?

wojtek-t replied:
I think it shouldn't

jpbetz replied:
Right, I think we can do this with the existing etcd functionality.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+# Consistent Reads from Cache
+
+Kubernetes Get and List requests are guaranteed to be "consistent reads" if the
+resourceVerions paramter is not provided and are served from etcd using a
```
nit: resourceVersion parameter

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+## Summary
+
+Consistent reads may be served from cache so long as:
+- A consistent (quorum) read is first made to etcd to get the latest "revsion"
```
nit: revision

wojtek-t replied:
same below

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+
+## Motivation
+
+Intuitively, we expect that serving reads from the watch cache to be more
```
It's not just intuitively - it is based on data and facts.
If serving from etcd, all data has to be send from etcd and deserialized in apiserver. Those steps doesn't exist when listing from watch cache.

jpbetz replied:
Absolutely. I've firmed this up.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
+
+Since etcd progress notify is only available in etcd 3.4. We would only serve consistent reads from cache if `--storage-backend` is set to `etcd3.4` or higher.
```
Re my comment above - that would mean we would have to parse this flag to detect it - won't introducing a dedicated bool flag be easier for it?

And more explicit?

jpbetz replied:
We'll need a feature gate for progress notify. I've clarified that in the doc a bit better. For alpha, the `WatchCacheConsistentReads` flag will be off by default. And once in beta, it would be on by default but still possible to disable by default.

And once the feature is beta, it's would only enabled if the the detected etcd version is new enough. I've rewritten the section about etcd version detection to match.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+Guard this optmization behind `WatchCacheConsistentReads` feature gate, and only server consistent reads from cache if it is enabled.
+
+When an consistent read request is received and the watch cache is enabled:
+- Use a range(limit=0) request to get the current revision from etcd for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
```
Maybe I'm completely wrong, but let me throw this idea:
- IIUC, the requests to etcd with limit=N set, result in forking the underlying tree to allow serving the "continuation" request later; so intuitively, if we will be sending a lot of those, that may be relatively expensive
- what if instead, we will be sending requests with `WithCountOnly` set? is computing count much more expensive or is it also computed in constant time (or close to constant?)

@jpbetz @jingyih

jingyih replied:
> IIUC, the requests to etcd with limit=N set, result in forking the underlying tree to allow serving the "continuation" request later; so intuitively, if we will be sending a lot of those, that may be relatively expensive

When limit=N set, etcd server reads N+1 keys from disk. 
@jpbetz limit=0 means no limit, so it is not what we want.

> what if instead, we will be sending requests with WithCountOnly set? 

This is a good idea. WithCountOnly is served from the in memory index tree in etcd - no need to read from disk.

We could do some experiments to measure which is the cheapest way to get the current revision from etcd.

jingyih replied:
> IIUC, the requests to etcd with limit=N set, result in forking the underlying tree to allow serving the "continuation" request later;

BTW, "continuation" is achieved by asking for the same revision in the next request.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful p ...
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
```
We may want to tune timeout for that case to avoid waiting for too long.

hormes replied:
In fact, we also need to convert etcd progress notify to a bookmark event. Otherwise, we may not wait for the revision.
https://github.com/kubernetes/kubernetes/pull/77584

hormes replied:
See this mentioned below.

wojtek-t replied:
This is more tricky - there is possibility to overload clients with bookmarks.
I agree we need to process them, but I'm not convinced that those should be translated directly to bookmarks.

hormes replied:
It will only affect the reflectors used to watch etcd in apiserver, not the client's watcher.

wojtek-t replied:
I don't fully agree.
This would be true if you always has watchcache enabled for everything. This isn't true for all setups as well as for all resources (e.g. events don't have watch cache enabled by default). So a watch connected to apiserver without watch cache is equivalent for watch to propagate watch cache.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful p ...
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
```
What I'm slightly worried about is the fact that it adds additional roundtrip to etcd, which may not be negligible in cases where etcd is running somewhere else.

wojtek-t replied:
To clarify - I'm not worried in the case of fallback - then additional latency sounds reasonable to me (because that means apiserver has some problems with watch cache).
What I am worried, is requesting this progress notify even explicitly.

hormes replied:
If we only add consistent reads to the `List`, we don't have to worry too much about it. There is not much problem in `Get` a single etcd key.

wojtek-t replied:
I don't think that fully solves the problem, because in theory there can be many lists for low-count objects, but given how the feature works it will result in sending bookmarks to every etcd watcher (also from other resources), which depending on the implementation may be even up to extremely expensive...

jpbetz replied:
I've split the proposal into two alternatives that we will compare and pick the better of. I think Alternative 2, which we're currently experimenting with, largely would address the main concern here. PTAL.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+Optional: In addition to progress notify, also configure etcd to send regular
```
@jingyih - how hard would be to build an experimental build (on top of 3.4.x) and expose that option as a flag?
Even if that is not something that can be cherrypicked to 3.4 branch, this option sounds like a better one medium term, so I would like to at least be able to evaluate them.

jingyih replied:
> how hard would be to build an experimental build (on top of 3.4.x) and expose that option as a flag?

Should be easy.

wojtek-t replied:
Jingyi - if you could help with that, I'm happy to help (myself or by finding someone) with experiments to validate its impact at scale.

jingyih replied:
https://github.com/etcd-io/etcd/pull/11463

The newly added server flag name is `--experimental-watch-progress-notify-interval`.

jpbetz replied:
Thanks for providing this @jingyih

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+- If the block times out, skip the cache and serve the request directly from storage.
+
+Optional: In addition to progress notify, also configure etcd to send regular
+progress events all on watches, every 10-100 milliseconds or so. If this is
```
100ms might be slightly too long, but yeah - we can figure that out separately

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+- Use a range(limit=0) request to get the current revision from etcd for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful p ...
```
When I'm looking into documentation, for me it looks that the notification is send to every single watcher that exists:
https://github.com/etcd-io/etcd/blob/master/clientv3/watch.go#L74

If that's the case, it's not exactly what we would want...
@jpbetz @jingyih

jpbetz replied:
It's all watch channels, but only for the client that requested it (not all clients). Let's check this carefully, but since we create an etcd client per storage backend, we only have one watch per etcd client and this might be fine?

wojtek-t replied:
I digged a bit deeper into it and I guess i understand it now:
- each watch can set the `WithProgressNotify` option
- progress notifies are send only to the watches that specify it (but to all of those)
- progress notifies are send periodically by the server (default 10m?)
- additionally you can request for sending it and then its send again to all watchers that requested it

It may be obvious to everyone except from me, but I think it would be useful to clarify it in this proposal too (maybe also in etcd documentation?)

hormes replied:
yes, the 1st~3nd are available in etcd 3.x, the 4th is available in etcd 3.4

jpbetz replied:
Thanks, I've added all this to the document.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability).
+
```
One more option @jpbetz and myself were talking about offline yesterday is the following:
- imagine that we introduce a new API to etcd that will be along those lines:
"Request: send my a response (containing id of this request) via watch once this request will be linearized and process via quorum"

This would allow us to send such request and simply serve response from watch cache on getting appropriate watch event.
It's obviously much more work on etcd side, but I would add that here as potential future improvements.

jpbetz replied:
Added to alternatives

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+request must be served from etcd anyway, which will be much higher latency than
+immediately serving from etcd like we do today. If this is an issue, it might be
+possible to partially mitigate it by tuning the “wait until fresh” duration to a
+sufficiently short value.
```
I think we will have to do that anyway.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,191 @@
+#### Reflector
+
+- Provide a way for reflectors to be configured to use resourceVersion=”” for initial list
+- Upgrade all in-tree reflectors to use resourceVersion=”” (TODO: this will be problematic for cluster that do not have this feature enabled)
```
We should evaluate it when we will be ready with the implementation.
But assuming we will hear a lot of feedback that people are not yet using etcd 3.4, if we would expose the option above (ability to configure reflector) via flag or some other configuration option on components, that would allow everyone to configure it.
And we can leave the default still set to "0", and switch it once this will appear to be much safer.

jpbetz replied:
Yes, this is the best available approach. I've added it to the doc.

mm4tt commented:
Interesting idea!

I took a cursory glance a question came to my mind. How would this work for paginated list calls? I couldn't find anything in the proposal on that.

wojtek-t commented:
> I took a cursory glance a question came to my mind. How would this work for paginated list calls? I couldn't find anything in the proposal on that.

It's not solved. That said, I think that we don't necessary need to address that from day1.
In my opinion, the biggest gain from it we can get from node-originating requests (e.g. kubelet listing pods scheduled on its node). For those requests, the size of the response is small (it fits a single page, assuming you won't make it extremely small), whereas the number of objects to process is proportional to cluster-size (so fairly big).

So as an example, in 5k-node cluster with 30pods/node, listing "pods from node X" returns 30 items, but listing from etcd would result in listing 150k pods, deserializing them, filtering and returning those 30. Listing from cache means just filtering and returning those 30, and once https://github.com/kubernetes/kubernetes/pull/85445 merges (and one more small PR on top of it), it will be proportional to the size of result.

So yes - this doesn't immediately solve all problem, but it may visibly help with many of those.

I can imagine the algorithm like:
- if limit is set, check if we set selectors - if not fallback to etcd, if so, assume high probability the result will not be too large
- ensure cache is up-to-date
- try to list from cache
- if result is too large, fallback to list from etcd

@jpbetz - while the above algorithm is probably not something we want to put in the KEP (do we?), the above explanation I made actually is probably worth putting (parts of it should do into motivation, I think).

detiber commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,252 @@
+### Goals
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scailability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
```
```suggestion
- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
```

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,252 @@
+response is small (it fits a single page, assuming you won't make it extremely
+small), whereas the number of objects to process is proportional to cluster-size
+(so fairly big). I.e. when kubelets requests pods, they request all pods and then
+use a selector to filter the pod list down to the ones scheduled on that kubelet.
```
I think this is slightly misleading - filtering is done in apiserver, but if the LIST is served from etcd, etcd returns all pods, and only apiserver does the filtering (which e.g. requires deserialization of all pods)

jpbetz replied:
I've made this more explicit, and used the 5k node example you mentioned.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,252 @@
+
+We will need to measure the impact to performance and scalability, but we have
+data and experience from prior improvements made by the watch cache that
+suggests there is significant opportunity here, and we would like to explore it.
```
nit: I'm 100% convinced that this will help performance/scalability - issues like https://github.com/kubernetes/kubernetes/issues/86483 actually already prove it. We just need to measure by how much.

jpbetz replied:
I agree. I've rephrased to match.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,252 @@
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough, we may not need to request progress notify events from etcd.
```
FYI: https://github.com/kubernetes/kubernetes/pull/86769 is the base for experiments

hormes replied:
This can only be considered experimental, the etcd kubernetes cluster used may serve other clients which maintains a large number of watchers, this value cannot be modified in a production environment.

jpbetz replied:
Our current goal is for a KEP is to describe what we need to take this feature to alpha (defaulted to off/experimental), so I think that's fine for now.

hormes commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,252 @@
+
+Because the watch cache does not paginate responses, clients that were
+previously getting pagination for RV="" requests will start getting unpaginated
+responses when this feature is enabled.
```
As I mentioned in [the comment](https://github.com/kubernetes/kubernetes/issues/86483#issuecomment-571404985), the most important point to consider is how to handle the page list

jpbetz replied:
I've added this to the graduation criteria so we don't go to beta until it is fully resolved.

I've also added a section about paging and have started to describe the problem. Let's continue to work on that and make sure we're satisfied with what we'd do for alpha before we mark this KEP implementable.

answer1991 commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,252 @@
+partially mitigated by configuring etcd to send regular progress events, which
+is something we will explore.
+
+In the worst case, the “wait until fresh” duration is exceeded, and then the
```
Should we provide an option like "RV=-1" which can skip "wait until fresh" and serving from etcd directly?

Because I think "wait until fresh" is not real "consistent read", as the data has been changed during the wait time.

hormes replied:
"Consistent read" here is similar to etcd quorum read, which in fact is exactly refers to linearizability consistency in distributed systems.

jpbetz replied:
+1, the consistency we're guaranteeing by waiting is exactly equivalent to the guarantee provided by linearizability consistency, so we can safely call it a consistent read. Note that in a distributed system there are many things that can result in a delay: network latency, gc pauses, etc.. Consistency isn't defined by how fast a response is returned, but by what properties it has.

wojtek-t replied:
+1

jpbetz commented:
re https://github.com/kubernetes/enhancements/pull/1404#issuecomment-566509864
> Joe Betz - while the above algorithm is probably not something we want to put in the KEP (do we?), the above explanation I made actually is probably worth putting (parts of it should do into motivation, I think).

This is very useful. I've added it to the motivation section.

For pagination I've added a small section about it and also listed it in the graduation criteria section as something we'd need to address fully before going to beta.

jpbetz commented:
> To clarify - I'm not worried in the case of fallback - then additional latency sounds reasonable to me (because that means apiserver has some problems with watch cache).
What I am worried, is requesting this progress notify even explicitly.

I'm a bit worried about that too.

I've restructured the doc to propose two alternatives: 1) Use WithProgressNotify to enable automatic watch updates, and 2) Use WatchProgressRequest to request watch updates when needed. If we can demonstrate alternative #1 is sufficient, I don't think we need to explicitly request progress notifies. Let's figure that out experimentally, trying alternative #1 first like you suggested.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,287 @@
+modify etcd to send them must more frequently, e.g. every 20ms.
+
+When an consistent read request is received and the watch cache is enabled:
+- Get the current revision from etcd using a range read with limit=0 for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
```
I think that @jingyih already mentioned that limit=0 won't work (because it serves everything).
My proposal was to return "WithCount", but we need to evaluate performance.

jpbetz replied:
You're right. limit=0 won't work. We could do a read for a guaranteed to be empty range with limit=1 for added safety, or we could do WithCount. I'll update the doc to summarize these options.

jpbetz commented:
> I don't think that fully solves the problem, because in theory there can be many lists for low-count objects, but given how the feature works it will result in sending bookmarks to every etcd watcher (also from other resources), which depending on the implementation may be even up to extremely expensive...

I've rewritten this sections to more clearly spell out the two alternatives we have, which hopefully makes more sense.

jpbetz commented:
@lavalamp @deads2k We'd get try this in 1.18. Would one of you have a look?

jpbetz commented:
/assign @lavalamp @deads2k

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+
+#### Alternative 2: Use WatchProgressRequest to request watch updates when needed
+
+etcd 3.4+ provides a `WatchProgressRequest` request that can be made on a watch channel. When requested,
```
What happens if we send a "WatchProgressRequest" to an etcd version that doesn't understand it?

jpbetz replied:
etcd responds with a `NOT_FOUND` (https://github.com/grpc/grpc/blob/master/doc/statuscodes.md) code.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+When an consistent read request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly (20ms?) for the watch to catch up to the current revision.
```
Why is 20ms the number you're anchoring everyone on? :)

jpbetz replied:
I like your suggestion about 250ms.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+
+#### Comparing the alternatives
+
+We will experiment with modifying etcd to automatically send progress events in
```
Something to consider, for the priority and fairness effort we are going to be letting less critical requests wait in queues. Which means by the time the request gets to the watch handler, it may have sat in a queue for N seconds, like as many as 15.
* Maybe 20ms is more aggressive than necessary?
* Maybe the data served only needs a guarantee that it's at least as recent as when the request arrived at apiserver?

Like, what if once or twice a second apiserver did a consistent read, and stores a mapping from time to RV. Then, when handling a request, it looks at the request's arrival time. If it's old enough that we've got a subsequent time:RV value, then we don't have to make a separate trip to etcd and probably the cache won't have to wait any additional time before serving.

Otherwise, we could make a special request of etcd to get the global index (as in the design as currently written), or we could just wait for the next one.

jpbetz replied:
> Maybe the data served only needs a guarantee that it's at least as recent as when the request arrived at apiserver?

Yes, this is certainly true. What we're specifically guaranteeing is linearizable consistency. An intuitive way to think about this is: If some agent performs some write and gets a response from etcd confirming that write, a subsequent read (either by that agent, or another agent that is told that write read as been completed) should return data at a state no older than that write. Delays before serving the read and delays returning the result are expected (being a distributed system and all) and don't impact correctness.

> Like, what if once or twice a second apiserver did a consistent read, and stores a mapping from time to RV. Then, when handling a request, it looks at the request's arrival time. If it's old enough that we've got a subsequent time:RV value, then we don't have to make a separate trip to etcd and probably the cache won't have to wait any additional time before serving.

Yes, this works. The happensBefore chain provides the exact same guarantee. When read concurrency is high, somehow minimizing the "what's the current RV" requests to etcd would make a lot of sense. They could either be periodic, or another why I had thought about doing it is to accumulate requests that are received over some interval, and at the interval make one "what's the current RV" request to etcd, and then serve all the requests from cache.

I'll add it to the potential optimizations.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+alternative 2.
+
+We will also explore how long the kube-apiserver should wait for a watch cache
+to catch up to the needed revision. I.e. for a progress event interval of 20ms,
```
If people want minimal latency, they can provide an RV, no?

I think it's pretty reasonable to make them wait an extra 500ms for consistent reads; so I think 250ms (to allow for a near miss) might be a more reasonable number than 20ms.

jpbetz replied:
Sounds good. @mm4tt ran some experiments (https://github.com/kubernetes/kubernetes/pull/86769) and a progress interval of 50ms did result in noticeable performance impact, so next we can do a run with 250ms and see how that performs.

lavalamp replied:
We could even go to 333ms, but not all the way to 500ms, I think--we wouldn't want 2 of those to miss. I think we should try and keep the overall latency below a second.

jpbetz replied:
250ms did fine in the last scalability test run. We'll go with that in the doc for now.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+etcd 3.x versions.
+
+Allow the etcd minor version to be specified in the kube-apiserver
+`--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  When client
```
I'd really prefer not to do this if at all possible.

jpbetz replied:
Agreed. It looks like we're starting to strongly favor alternative 1, where we don't need to send etcd `WatchProgressRequest`s at all. I'll remove alternative 2 from this KEP and remove this section as well.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+there might be an etcd version mismatch and fallback to the etcd functionality
+supported by all etcd 3.x versions.
+
+Also, Due to etcd upgrades and downgrades, there is no way to automatically
```
Can you expand this? Why?

jpbetz replied:
etcd distinguishes between the version of the binary a etcd member in the cluster is running and the "cluster" version. At any time client can be connected to an etcd member the cluster version can change, either as part of a cluster upgrade or downgrade. If a client checked the version of the cluster when it established a connection, there is potential for it to change. Even if the client checked the version of the cluster before making an API call, the cluster version could change before the API call is served.

lavalamp replied:
That argues for just trying it and seeing if it works, if we wanted to do this?

jpbetz replied:
This would only matter for the design alternative we just moved to rejected, but the idea was that if went this route, we'd need to rely on administrators setting the etcd version via flag correctly to 'etcd-3.4' only when they're certain that they're running an etcd that will stay at that cluster version or above, since we cannot have etcd automatically detect the version in a way that is guaranteed to be correct.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+previously getting pagination for RV="" requests will start getting unpaginated
+responses when this feature is enabled.
+
+The impacts of this can be explored when the feature is in alpha.
```
I think this doc needs at least a sketch of a plan for dealing with this. Like, this sounds like a showstopper for beta and up, so we need to know that it's solvable or the design is a dead end.

liggitt replied:
I definitely agree with that.

jpbetz replied:
Sounds good. I'll explore options and expand this section.

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,294 @@
+### Test Plan
+
+We will need to scale and performance test this carefully. We’ll need to measure
+latency, throughput, volume of progress notify requests, and change to volume of
```
I think this section should state how exactly you'll evaluate this? Measuring these things seems necessary but not sufficient?

Is this a correctness change or a performance change? Both?

lavalamp replied:
Like, I'm willing to pay latency to get throughput on startup + correctness. It'd be good to document how much we're willing to pay for how much benefit.

jpbetz replied:
Initially we're improving performance, primarily of LIST (with no RV provided), that doesn't change correctness. But then after that we're going to use instead of LIST (with RV=0) to improve correctness. We do expect a latency increase on those LIST calls but other wise no net loss to scale. I'll clarify with more detail in the doc.

@mm4tt and @wojtek-t have done some scale tests on etcd progress notify. I'll include more details here as well as what else we need to test.

lavalamp commented:
Well, I'd just move it to a "rejected alternatives" section and add
reasons, rather than delete it completely.

On Fri, Jan 24, 2020 at 12:41 PM Joe Betz <notifications@github.com> wrote:

> *@jpbetz* commented on this pull request.
> ------------------------------
>
> In keps/sig-api-machinery/20191210-consistent-reads-from-cache.md
> <https://github.com/kubernetes/enhancements/pull/1404#discussion_r370832815>
> :
>
> > +Optional: For some (but not all) of the etcd progress watch events, also create a
> +kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
> +reflectors and shared informers are kept up-to-date. The benefit of this is that
> +it minimizes the chance that these clients will end up with an out-of-date
> +resource version and need to relist (which can impact scalability). See [Watch
> +Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
> +for details.
> +
> +### Make the kube-apiserver aware of etcd's minor version
> +
> +This is only needed if we go with Alternative 2 (Use WatchProgressRequest to
> +request watch updates when needed) since Alterative 1 is compatible with all
> +etcd 3.x versions.
> +
> +Allow the etcd minor version to be specified in the kube-apiserver
> +`--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  When client
>
> Agreed. It looks like we're starting to strongly favor alternative 1,
> where we don't need to send etcd WatchProgressRequests at all. I'll
> remove alternative 2 from this KEP and remove this section as well.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/enhancements/pull/1404?email_source=notifications&email_token=AAE6BFTRIUOEXQGPAW2RCJ3Q7NG6LA5CNFSM4JZHHM5KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCTAUXHA#discussion_r370832815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFTHD2Z2PQPG6Q6WOA3Q7NG6LANCNFSM4JZHHM5A>
> .
>

jpbetz commented:
Thanks @lavalamp @liggitt. Feedback applied. Added a detailed review of options for addressing pagination issue and summarized the main correctness/performance/scalability testing we'd need to do.

wojtek-t commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,404 @@
+  the resource versions it has returned continuation tokens to clients for.
+
+Both of these are major changes, and would require scalability validation. We
+are not planning to pursue this option.
```
Actually, I was thinking about that and maybe we can do some poor-man support for pagination.
Consider the following:
- watch cache is getting LIST request with pagination
- we list everything from the internal cache and have pointers to those objects
- we return first LIMIT of those, and in the internal map store the remaining ones indexed by "continuation token" that we just generated
- we GC the items from this map after N seconds from insertion
- if continuation is set in the request, we lookup that map and return next LIMIT items, if the item doesn't exist in the map we (either fallback to etcd or return an error - probably the former)

These kind of changes should be fairly simple and maybe those actually be enough?

@lavalamp @liggitt @jpbetz - for thoughts

tedyu replied:
Determining good value for N (seconds after which items are garbage collected) may be tricky.
Another measure is to bound the memory consumed by this map.

jpbetz replied:
Something like this could work. We'd need to think through out the memory utilization/bounds. I'll add a note and we can expand on it.

mm4tt commented:
FYI, I'm re-running the tests with 250ms interval to see how much it helps (comparing to 50ms interval). Will post the results once I have them (most likely tomorrow).

lavalamp commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,404 @@
+that a progress notify interval of 50ms results in a noticeable increase in both
+etcd and kube-apiserver CPU utilization and some increased kube-apiserver serving
+latency. We intend to address this by first trying a 250ms progress notify interval,
+if ths performs well we will use that as our interval. If not, we will dig in
```
> don't buy @lavalamp `s arguments that with Priority and Fairness we will have many requests waiting in queues for long times. In some cases for very overloaded clusters - sure

The argument is basically:
* Right now, everyone leaves a ton of extra resources for the control plane, because if you exceed your limits, the cluster more or less becomes unusable.
* Once that is no longer true, people will size their control plane to match their usage (for cost reasons if no other)
* Maximum efficiency implies non-zero queue lengths
* Many of our requests are < 100ms
* Queue length can be ~10 without breaking our SLO very often
* Non-important requests don't need anything like our SLO, for those, the queue length can be much longer

Basically, it will make all clusters into "heavily loaded" ones, since it will make heavily loaded clusters usable.

jpbetz replied:
Added a short section about the reasoning for going with 250ms.

wojtek-t replied:
@lavalamp - I don't agree with the argument. Mostly with this one:

> Once that is no longer true, people will size their control plane to match their usage (for cost reasons if no other)

From scalability POV, I will never say that people should now allocated less resources for control plane. What the priority & fairness will give us imho is not the ability to save a bit on control plane resources. but rather the fact that the cluster won't completely blow up when either (a) some spike in load or (b) something bad causing cascading failure will happen in the cluster.

I can imagine that some people will decrease amount of resources for control plane for small clusters (where those are significant percentage). But for the largest ones (which is where this will help most and those where we can potentially have biggest issues), I bet it won't happen (as the gain from it is also close to negligible).

xiang90 replied:
> But for the largest ones (which is where this will help most and those where we can potentially have biggest issues), I bet it won't happen (as the gain from it is also close to negligible).

I guess I am one of the people who run the largest clusters. We run control plane on beefy machines. Even with priority/fairness support, more efficient caching etc., it will not change. The resource overhead for a large cluster is negligible. Over provisioning gives us the flexibility and time to react for unexpected problems.

wojtek-t replied:
Yeah - that's exactly what I had on my mind writing my above comment. Thanks @xiang90

jpbetz commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,404 @@
+current revision" requests that need to be made to etcd in exchange for higher
+request latency (but only for consistent reads). A simple implementation would
+be to do this on a fix interval, where, obviously, if there were not reqests
+during the period, we don't bother to fetch a current revision from etcd.
```
I'll demote this to the "future improvements" section and put a comment about the potential downsides too.

jpbetz commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,404 @@
+To get the revsion we have some options: 
+
+- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision 
+- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
```
Agreed. I've clarified.

jpbetz commented:
Feedback from latest comments applied.

As a reminder, 1.18 KEPs are due today. Should we try to get this one in?

mm4tt commented:
> FYI, I'm re-running the tests with 250ms interval to see how much it helps (comparing to 50ms interval). Will post the results once I have them (most likely tomorrow).

I documented the results at https://github.com/kubernetes/kubernetes/pull/86769#issuecomment-579171765 to keep everything in one place. TL;DR it looks good from the scale/perf perspective.

anguslees commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,445 @@
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
+- If the block times out, skip the cache and serve the request directly from storage (etcd).
```
This seems like a cascading failure waiting to happen :(  I think this fallback step could be rephrased as "If the cluster appears to be struggling, then do a more expensive operation."

If it's ok to do this fallback, then I counter-propose we just skip the cache for _all_ consistent reads.  At least in this case the apiserver/etcd cluster will be sized correctly, since the read load will be continuously apparent.

More constructively ;) I suggest we change this fallback step to "If the block times out, return GOAWAY", with a big comment explaining why we don't fallback to reading directly. Or alternatively (better?), just make `waitUntilFreshAndBlock` block forever (ie: until the client's own timeout gives up and retries/aborts).

wojtek-t replied:
I don't think "goaway" is good enough proposal - I think that what you're just saying is that we need additional layer of rate-limitting at this point. Which is a fair point.

jpbetz replied:
You're right, we don't want to fallback to etcd. I've updated the doc and added a section mentioning both "Retry-After" and priority and fairness as options for dealing with rate-limiting clients.

liggitt replied:
Revisiting this thread in light of https://github.com/kubernetes/kubernetes/pull/126139 / 

Struggling etcd *could* be *one* reason the block timed out, but it is not the only reason. *Anything* that makes the watch cache laggy will cause this failure mode, and there are known causes apart from struggling etcd that trigger that.

As a first step to ensure this feature doesn't break existing working users in new ways, @deads2k and I agree that we *should* fall back to serving from etcd (the existing code path) if the list-from-cache times out. If we instrument when that occurs, and instrument when the fallback succeeds, we'll get visibility to users and clusters that would be broken by defaulting this on with no fallback.

anguslees commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,445 @@
+We will explore how long the kube-apiserver should wait for a watch cache to
+catch up to the needed revision. I.e. for our chosen progress event interval,
+should the kube-apiserver also wait a similar interval for the required revision
+to become available or should it fallback to making the request to etcd?
```
See above comment about cascading failures and the need for load shedding here, not load amplification.

jpbetz replied:
Thanks, I've updated this section as well.

jpbetz commented:
@lavalamp, @deads2k Can we merge this PR as provisional? @jfbai has proposed an improvement to this KEP in https://github.com/kubernetes/kubernetes/pull/88264 and I'd like to get it merged so we can more easily collaborate. All outstanding comments on this PR have been addressed.

lavalamp commented:
I think I'm OK with provisional. @deads2k?

jfbai commented:
> @lavalamp, @deads2k Can we merge this PR as provisional? @jfbai has proposed an improvement to this KEP in [kubernetes/kubernetes#88264](https://github.com/kubernetes/kubernetes/pull/88264) and I'd like to get it merged so we can more easily collaborate. All outstanding comments on this PR have been addressed.

Sure, I will update the idea using `WithLastRev` into this KEP after it is merged.

jfbai commented:
> @lavalamp, @deads2k Can we merge this PR as provisional? @jfbai has proposed an improvement to this KEP in [kubernetes/kubernetes#88264](https://github.com/kubernetes/kubernetes/pull/88264) and I'd like to get it merged so we can more easily collaborate. All outstanding comments on this PR have been addressed.

Sure, I will update the idea using `WithLastRev` when after this KEP is merged.

deads2k commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,455 @@
+Not all LIST requests require consistency, and we're taking the view that if you
+really want to have a consistent LIST (we should explicitly exclude GETs from
+it), then you may need to pay additional tax latency for it. For this reason, we
+intend to start with a 250ms progress notify interval, which will on average 125ms
```
This tax seems pretty significant.  Please also mention it in a more prominent section.

jpbetz replied:
I'll highlight it. Also, @jfbai's proposal for using the etcd `WithLastRev` request option might eliminate this tax. We'll open a follow up PR to explore that once this is merged provisional.

deads2k commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,455 @@
+intend to start with a 250ms progress notify interval, which will on average 125ms
+latency to each consistent LIST request.
+
+The requests this is expected to impact are:
```
Doesn't this also impact `kubectl get`?

jpbetz replied:
`kubectl get` uses `limit=500` so the section about pagination applies to it; current plan is to not do consistent reads from cache for paginated requests. I'll clarify.

deads2k commented:
I don't see a section talking about how I get to opt out of this behavior and still get a "normal" quorum read.  Given the difficulty we've had with nearly every watch cache feature, I think we'll need that ability for our own debugging if nothing else. When these problems struck in the past, we were able to do find an affected cluster and use a mix of cached and live requests to figure out what was happening.  I don't want to give up that ability entirely.

I don't think this is a case where, "then everyone will use it", because the normal flow should be good enough to convince people to use it.  If it's not and we truly want to force them, it could become an ACL check, but hopefully the feature is good enough that everyone wants to use it.

deads2k commented:
Add the "UNRESOLVED" markers on the sections I commented on (https://github.com/kubernetes/enhancements/pull/1545/files#diff-bb6317ae71eb96981343e559d9598d80R38) and I can go for provisional

deads2k commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,455 @@
+
+We are not planning to pursue this option.
+
+### What if the watch cache is stale?
```
summarize this as a one liner in the top level details: "any failure for the cache to come 'live' will result in API request rejection"

deads2k commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,455 @@
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
```
if you aren't going to try to avoid allowing true quorum reads as we have them today, please list as an explicit non-goal.  Also, per my overall comment, I think this is a shortcoming and should be addressed.  it doesn't need to be a default, but it should be possible.

jpbetz replied:
I think this needs more discussion. I've added a unresolved in non-goals.

jpbetz commented:
Thanks @deads2k. Feedback is applied and `<<[UNRESOLVED]>>...<<[/UNRESOLVED]>>` sections have been added.

wojtek-t commented:
I would like to make a one more deeper pass on it before we move it to implementable and I won't have time for it in the next 2.5 weeks, but I'm definitely fine with provisional.

jpbetz commented:
> I would like to make a one more deeper pass on it before we move it to implementable and I won't have time for it in the next 2.5 weeks, but I'm definitely fine with provisional.

Sounds good, I've listed @wojtek-t, @lavalamp and @deads2k as the approvers for this to make sure we get your feedback.

jpbetz commented:
@deads2k, @lavalamp Would one of you tag this?

xiang90 commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,478 @@
+
+Create etcd watches with `WithProgressNotify` enabled (available in all etcd 3.x versions).
+
+When `WithProgressNotify` is enabled on an etcd watch, etcd sends progress
```
the proposal to make it configurable on a per watch stream basis.

xiang90 commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,478 @@
+
+When an consistent LIST request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
```
What does "current revision" mean exactly? We can get the largest modified revision for the listed resources (vs current revision of the entire key space). This can reduce the block time (for less frequently changed objects, there should be just no blocking).

Alternatively, we can add a new watch request type "urgent" or something to let etcd simply deliver all pending watch events or an empty result (if there is no new changes).

xiang90 commented on keps/sig-api-machinery/20191210-consistent-reads-from-cache.md:
```diff
@@ -0,0 +1,478 @@
+### Non-Goals
+
+<<[UNRESOLVED @deads]>>
+- Avoid allowing true quorum reads. We should think carefully about this, see: https://github.com/kubernetes/enhancements/pull/1404#discussion_r381528406
```
i feel the "right and traditional" solution for this is to enable v system leasing in the apiserver etcd pkg. here is the description on how it works: http://web.stanford.edu/class/cs240/readings/89-leases.pdf. we already implemented it here: https://github.com/etcd-io/etcd/pull/8341/files. we can port the code into the apiserver side if we want. but this is a pretty heavy :P

lavalamp replied:
Why would we stop permitting true quorum reads? I didn't get that at all out of this doc when I read it?

deads2k commented:
/lgtm

I think this is a good point to start iterating from. The unresolved sections will make sure we dont' accidentally skip agreeing on what's in and out.
